Add a section to help users understand key SDK and core concepts with samples.

ahsonkhan opened the pull request:
Addresses https://github.com/Azure/azure-sdk-for-cpp/issues/1084#issuecomment-795806727

antkmsft commented on README.md:
```diff
@@ -66,20 +66,25 @@ In order to use the SDK installed via vcpkg with CMake, you can use the toolchai
-All the Azure C++ SDK headers needed to be included are located within the `<azure>` folder, with sub-folders corresponding to each service. Similarly, all types and APIs can be found within the `Azure::` namespace. For example, to use functionality form `Azure::Core`, include the following header at the beginning of your application `#include <azure/core.hpp>`.
+The **entry point** for most scenarios when using the SDK will be a top-level client type corresponding to the Azure service you want to talk to. For example, sending requests to blob storage can be done via the `Azure::Storage::Blobs::BlobClient` API. All APIs on the client type send HTTP requests to the cloud service and return back an HTTP `Response<T>`.
+
+All the Azure C++ SDK headers needed to be included are located within the `<azure>` folder, with sub-folders corresponding to each service. Similarly, all types and APIs can be found within the `Azure::` namespace. For example, to use functionality from `Azure::Core`, include the following header at the beginning of your application `#include <azure/core.hpp>`.
```
I don'tthink users will need to include core that often in real world. Should we use <azure/storage/blobs.hpp> as example instead?

ahsonkhan replied:
I don't think service client library headers include all of core.hpp, I think adding core.hpp is likely going to be useful for end users for using things that might be used by one service but not another.

antkmsft commented on README.md:
```diff
@@ -66,20 +66,25 @@ In order to use the SDK installed via vcpkg with CMake, you can use the toolchai
 #include <iostream>
+
+// Include the necessary SDK headers
 #include <azure/core.hpp>
```
This is very likely not needed. Blobs should include everything they need from core.

antkmsft commented on README.md:
```diff
@@ -66,20 +66,25 @@ In order to use the SDK installed via vcpkg with CMake, you can use the toolchai
 using namespace Azure::Storage::Blobs;
 
-// Get the required connection string/key from an environment variable or Azure Key Vault.
+// Get the required connection string/key from an environment variable or Azure KeyVault.
```
I think the branding is "Azure Key Vault"

vhvb1989 replied:
correct, 
```suggestion
// Get the required connection string/key from an environment variable or Azure Key Vault.

```

vhvb1989 replied:
Actually, why are we mentioning Key Vault at all?  The sample is just for Storage isn't it?

I would remove the entire sample code from here and leave it only inside each service Readme to sample the code.  Maybe just add links to where the samples for the current services are.

ahsonkhan replied:
That note/comment is just there on where/how to read the connection string, to indicate best practices on where to get that from, the sample doesn't have anything to do with key vault.

JeffreyRichter commented on README.md:
```diff
@@ -66,20 +66,25 @@ In order to use the SDK installed via vcpkg with CMake, you can use the toolchai
 
-// Get the required connection string/key from an environment variable or Azure Key Vault.
+// Get the required connection string/key from an environment variable or Azure KeyVault.
 std::string GetConnectionString();
```
Is this a TODO or something?
We actually discourage the use of connection strings; can we change this?

ahsonkhan replied:
@JeffreyRichter how do you recommend we construct a blob client, if not using a connection string, as a copy-pasteable sample for customers?

vhvb1989 replied:
> We actually discourage the use of connection strings

Why?  For Storage, that's super easy to get the connection string from the Azure portal and use it.  Is it even easier to set the shareKeyCredential?

JeffreyRichter replied:
Connection strings usually combine multiple things together that should not be combined like endpoint, username & password. We want to encourage people to roll over passwords periodically while their service keeps running but if they are in a connection string, then they need a whole new connection string and there is no way to atomically update an endpoint, username, and password.

In addition, there are many reasons why distributed systems fail including bad endpoints and bad credentials. So, customers can better diagnose issues if they are responsible for passing these things explicitly – it is the cost of entry for building a distributed system.

Connection strings also have the potential for lots of problems like are they case sensitive? What does it mean if multiple credentials appear in the string? Will customers embed connection strings in their source code and check that in to github? (connections strings make this bad practice too easy).

Azure SDKs policy is to NOT support connection strings unless the portal/tooling does offer them. In this case, we support them for "get up and running fast" scenario but we still discourage their use for real production.

It's even in our C++ guidelines: https://azure.github.io/azure-sdk/cpp_introduction.html#cpp-authentication

antkmsft commented on README.md:
```diff
@@ -89,19 +94,96 @@ int main()
+    Azure::Response<Models::UploadBlockBlobResult> response = block.Upload(stream);
+
+    Models::UploadBlockBlobResult model = response.Value;
+    printf("Last modified date of uploaded blob: %s\n", model.LastModified.ToString().c_str());
```
Let's use `std::cout`, and then we also don't need the `c_str()`.

JeffreyRichter commented on README.md:
```diff
@@ -89,19 +94,96 @@ int main()
+    // for your scenario, such as BlobContainerClient.
+    BlobClient client
+        = BlobClient::CreateFromConnectionString(GetConnectionString(), containerName, blobName);
+    BlockBlobClient block = client.AsBlockBlobClient();
```
If you get rid of connection strings (which you should), then you should just new up a BlockBlobClient and not have to call AsBlockBlobClient.

JeffreyRichter replied:
I also suggest renaming the variable to blockBlobClient; not just block.

JeffreyRichter replied:
You could use SharedKeyCredential instead of connection string

antkmsft commented on README.md:
```diff
@@ -89,19 +94,96 @@ int main()
   }
   catch (const Azure::Core::RequestFailedException& e)
   {
+    printf("Status Code: %d, Reason Phrase: %s\n", (int)e.StatusCode, e.ReasonPhrase.c_str());
```
and also, we do mix and match `std::cout` and `printf`, and also we call `printf()` not as `std:printf()` - I don't like it.
Btw, do our headers do `using namespace std` or `include <stdio.h>` instead of `include <cstdio>` so that our code sample can call `printf()` withot an `std::`? That should not be happening.

antkmsft replied:
Also, this statement uses a C-style cast. Should be `static_cast<int>(e.StatusCode)`.

JeffreyRichter commented on README.md:
```diff
@@ -89,19 +94,96 @@ int main()
+- Exceptions for reporting errors from service requests in a consistent fashion via the base exception type `RequestFailedException`.
+- Abstractions for Azure SDK credentials (`TokenCredential`).
+- Handling streaming data and input/output (I/O) via `BodyStream` along with its derived types.
+- Polling long-running operations (LROs), via `Operation<T>`.
```
Add our paging support too?

ahsonkhan replied:
Yes! Good call out, we did finally add that.

antkmsft commented on README.md:
```diff
@@ -89,19 +94,96 @@ int main()
-Understanding the key concepts from the `azure-core-cpp` library, which is leveraged by all client libraries will also be helpful in getting started, regardless of which Azure service you want to use. You can find more information about them, with sample code snippets, here: https://github.com/Azure/azure-sdk-for-cpp/tree/master/sdk/core/azure-core#key-concepts
+#### Key Core concepts
+
+Understanding the key concepts from the `azure-core-cpp` library, which is leveraged by all client libraries will also be helpful in getting started, regardless of which Azure service you want to use.
```
Should we call the library `azure-core`, or even "Azure Core"? `azure-core-cpp` is vcpkg-specific package name under which we are available there.

JeffreyRichter commented on README.md:
```diff
@@ -89,19 +94,96 @@ int main()
+  std::string containerName = "testcontainer";
+
+  BlobContainerClient containerClient
+      = BlobContainerClient::CreateFromConnectionString(GetConnectionString(), containerName);
```
Replace connection string part of this example

antkmsft commented on README.md:
```diff
@@ -89,19 +94,96 @@ int main()
+
+Understanding the key concepts from the `azure-core-cpp` library, which is leveraged by all client libraries will also be helpful in getting started, regardless of which Azure service you want to use.
+
+The main shared concepts of `Azure::Core` include:
```
And here we call it `Azure::Core`. Let's just use "Azure Core"? Not `azure-core`, not `azure-core-cpp`, not `Azure.Core`.

JeffreyRichter commented on README.md:
```diff
@@ -89,19 +94,96 @@ int main()
+
+  // You can extract the T, from the returned Response<T>,
+  // which is typically named with a Result suffix in the type name.
+  Models::CreateBlobContainerResult result = createResult.Value;
```
Here the variable is 'result' but earlier the variable is 'model' - can we make consistent.

antkmsft commented on README.md:
```diff
@@ -89,19 +94,96 @@ int main()
+  Models::CreateBlobContainerResult result = createResult.Value;
+
+  // Now you can look at API specific members on the result object that is returned.
+  printf("Did the container get created? %d\n", result.Created);
```
std::cout

ahsonkhan replied:
Do you know what's the best way to emit a bool Anton?
Can you please share a suggestion/code snippet.

antkmsft replied:
`std::cout << "Did the container get created? " << std::boolalpha << result.Created << std::endl;`?
https://onlinegdb.com/SFgGxu_S-7

But I would not bother to put it, if C++ don't have it by default, plus applying `std::boolalpha` changes the behavior in all other statements in the program. Unless you set and restore ios flags, but all that diverts the attention. I'd say just do "`std::cout << "Did the container get created? " << result.Created << std::endl;`" - it is pretty good to illustrate the SDK without diverting attention.

antkmsft reviewed (approved):
LGTM given that you take in the feedback.

JeffreyRichter commented on README.md:
```diff
@@ -89,19 +94,96 @@ int main()
+  Models::CreateBlobContainerResult result = createResult.Value;
+
+  // Now you can look at API specific members on the result object that is returned.
+  printf("Did the container get created? %d\n", result.Created);
```
Perhaps CreateIfNotExists is a strange choice to demo. If creates it if it doesn't exist and doesn't if it does. So, unless it throws, the container does exist. So it is created which makes the printf text odd. I think a different function would be a better demo.

JeffreyRichter commented on README.md:
````diff
@@ -89,19 +94,96 @@ int main()
+You can intermittently poll whether the operation has finished by using the `Poll()` method on the returned `Operation<T>` and track progress of the operation using `Value()`. Alternatively, if you just want to wait until the operation completes, you can use `PollUntilDone()`.
+
+```C++
+  SomeServiceClient client;
````
I think you need to say that how this is initialized is not shown here otherwise a copy/past of this code is guaranteed to fail.

ahsonkhan replied:
Agreed, do you think I should use a concrete/real example, like start copy from uri instead?

JeffreyRichter replied:
Yes

JeffreyRichter commented on README.md:
```diff
@@ -89,19 +94,96 @@ int main()
+
+  while (!operation.IsDone())
+  {
+    Azure::Core::Http::RawResponse pollResponse = operation.Poll();
```
why not use auto here to make the code less pedantic

JeffreyRichter replied:
Also, I think it better to show PollUntilDone without the loop as this is the more common case and the code is much shorter.

ahsonkhan replied:
> why not use auto here to make the code less pedantic

I could do that, I was hoping to show more of the types/namespaces, but given `RawResponse` isn't something end users will care about right out the gate, I will use auto. Do you recommend I do that in a lot more places (auto everything) or just where it is not as relevant?

JeffreyRichter replied:
I think showing that the code is simple is important; so yes, I think auto everywhere is good.

antkmsft commented:
Would this take care of #2295, or is it different?

antkmsft commented:
> Would this take care of #2295, or is it different?

Ah, I see, this is for root readme, and the work item is about core readme, so probably different.

antkmsft commented on README.md:
```diff
@@ -66,42 +66,121 @@ In order to use the SDK installed via vcpkg with CMake, you can use the toolchai
-  std::string blobName = "sample-blob";
+  std::string endpointUrl = "<update to use your storage account url here>";
+  std::string accountName = GetAccountName();
+  std::string accountKey = GetAccountKey();
```
Just a proposal: implement `GetStorageAccountUrl()`, `GetAccountName()`, `GetAccountKey()` as one-liners returning `{ return std::getenv("AZURE_STORAGE_ACCOUNT_URL"); }`, and so on.

ahsonkhan replied:
I sort of agree, but the sample is already a bit verbose, this is mainly getting started, and the main intention here is to highlight the basic concepts.

I would probably also expect the user to add error handling when the environment variable doesn't exist. The user has to modify the code to choose their own environment variable key name anyway.

ahsonkhan replied:
Alright, took your suggestion Anton, thanks!

RickWinter commented on README.md:
```diff
@@ -66,42 +66,122 @@ In order to use the SDK installed via vcpkg with CMake, you can use the toolchai
+#### Using the SDK within your Application
 
-All the Azure C++ SDK headers needed to be included are located within the `<azure>` folder, with sub-folders corresponding to each service. Similarly, all types and APIs can be found within the `Azure::` namespace. For example, to use functionality form `Azure::Core`, include the following header at the beginning of your application `#include <azure/core.hpp>`.
+The **entry point** for most scenarios when using the SDK will be a top-level client type corresponding to the Azure service you want to talk to. For example, sending requests to blob storage can be done via the `Azure::Storage::Blobs::BlobClient` API. All APIs on the client type send HTTP requests to the cloud service and return back an HTTP `Response<T>`.
```
Nit, drop this
```suggestion
The **entry point** for most scenarios when using the SDK will be a top-level client type corresponding to the Azure service. For example, sending requests to blob storage can be done via the `Azure::Storage::Blobs::BlobClient` API. All APIs on the client type send HTTP requests to the cloud service and return back an HTTP `Response<T>`.

```

RickWinter commented on README.md:
```diff
@@ -66,42 +66,122 @@ In order to use the SDK installed via vcpkg with CMake, you can use the toolchai
-All the Azure C++ SDK headers needed to be included are located within the `<azure>` folder, with sub-folders corresponding to each service. Similarly, all types and APIs can be found within the `Azure::` namespace. For example, to use functionality form `Azure::Core`, include the following header at the beginning of your application `#include <azure/core.hpp>`.
+The **entry point** for most scenarios when using the SDK will be a top-level client type corresponding to the Azure service you want to talk to. For example, sending requests to blob storage can be done via the `Azure::Storage::Blobs::BlobClient` API. All APIs on the client type send HTTP requests to the cloud service and return back an HTTP `Response<T>`.
+
+All the Azure C++ SDK headers needed to be included are located within the `<azure>` folder, with sub-folders corresponding to each service. Similarly, all types and APIs can be found within the `Azure::` namespace. For example, to use functionality from `Azure::Core`, include the following header at the beginning of your application `#include <azure/core.hpp>`.
```
Nit, 
```suggestion
Azure C++ SDK headers are located within the `<azure>` folder, with sub-folders corresponding to each service. Similarly, all types and APIs can be found within the `Azure::` namespace. For example, to use functionality from `Azure::Core`, include the following header at the beginning of your application `#include <azure/core.hpp>`.

```

RickWinter commented on README.md:
```diff
@@ -66,42 +66,122 @@ In order to use the SDK installed via vcpkg with CMake, you can use the toolchai
 #include <azure/core.hpp>
 #include <azure/storage/blobs.hpp>
 
+// Use the appropriate namespace using directives
```
Nit
```suggestion
// Add appropriate using namespace directives

```

antkmsft reviewed (approved):
👍

antkmsft reviewed (commented):
Other feedback would be to maybe use `auto` more often, but it is subjective.

RickWinter commented on README.md:
```diff
@@ -66,42 +66,122 @@ In order to use the SDK installed via vcpkg with CMake, you can use the toolchai
 
-// Get the required connection string/key from an environment variable or Azure Key Vault.
-std::string GetConnectionString();
+// Since they are secrets, get the required account name and key string from an environment variable
```
Check the formatting before accepting this

```suggestion
// Secrets should be stored & retrieved from secure locations such as Azure::KeyVault.  For convenience and brevity of samples, the secrets are retrieved from environment variables.
//  Get the required account name and key string from an environment variable

```

vhvb1989 commented on README.md:
```diff
@@ -66,42 +66,122 @@ In order to use the SDK installed via vcpkg with CMake, you can use the toolchai
 {
-  std::string containerName = "testcontainer";
-  std::string blobName = "sample-blob";
+  std::string endpointUrl = "<update to use your storage account url here>";
```
```suggestion
  std::string endpointUrl = GetEndPointUrl();

```

antkmsft replied:
Do not break the word, use `Endpoint` instead of `EndPoint`

RickWinter commented on README.md:
```diff
@@ -66,42 +66,122 @@ In order to use the SDK installed via vcpkg with CMake, you can use the toolchai
   {
-    auto client = BlobClient::CreateFromConnectionString(
-        GetConnectionString(), containerName, blobName);
+    // There are many types of client types, use the one that's appropriate
```
I don't find this comment very helpful.  I suggest removing.

antkmsft replied:
BTW, should we keep it, I think it is better to rephrase as "there are many types of clients"

ahsonkhan commented:
> Other feedback would be to maybe use `auto` more often, but it is subjective.

I used it where it was redundant/obvious from the right hand side what the type was.
In other places, like the ones talking about `Response<T>`, the whole point of such samples is to showcase/teach the templated type to the user. Hiding it with auto, would make the coding pattern less visible.

antkmsft commented on README.md:
````diff
@@ -66,42 +66,122 @@ In order to use the SDK installed via vcpkg with CMake, you can use the toolchai
+You can intermittently poll whether the operation has finished by using the `Poll()` method inside a loop on the returned `Operation<T>` and track progress of the operation using `Value()`, while the operation is not done (using `IsDone()`). Your per-polling custom logic can go in that loop, such as logging progress.Alternatively, if you just want to wait until the operation completes, you can us ...
+
+```C++
+  std::string sourceUri = "<a uri to the source blob to copy>";
````
nit: mismatch between url (above) and url (here) usage. Also, maybe capitalize?

ahsonkhan replied:
I know. The API is called StartCopyFromUri, so that's where the inconsistency issue comes from :(

antkmsft replied:
I see, the API is called ...FormUri, it is better to align with the API then.

RickWinter commented on README.md:
```diff
@@ -66,42 +66,122 @@ In order to use the SDK installed via vcpkg with CMake, you can use the toolchai
+
+    auto blockBlobClient = BlockBlobClient(endpointUrl, sharedKeyCredential);
+
+    // Assuming there is some Azure::Core::IO::BodyStream containing the data to upload.
```
No assumption here, it has some random data that we are uploading.  I'd state "Create some data to upload into the blob"

vhvb1989 commented:
nit. Add links as `see more` or something from the samples to the package ReadMe.  For example, from the code snipped to the storage samples.

vhvb1989 commented:
/check-enforcer override

RickWinter commented on README.md:
```diff
@@ -66,42 +66,122 @@ In order to use the SDK installed via vcpkg with CMake, you can use the toolchai
-Understanding the key concepts from the `azure-core-cpp` library, which is leveraged by all client libraries will also be helpful in getting started, regardless of which Azure service you want to use. You can find more information about them, with sample code snippets, here: https://github.com/Azure/azure-sdk-for-cpp/tree/master/sdk/core/azure-core#key-concepts
+#### Key Core concepts
+
+Understanding the key concepts from the `Azure Core` library, which is leveraged by all client libraries will also be helpful in getting started, regardless of which Azure service you want to use.
```
```suggestion
Understanding the key concepts from the `Azure Core` library, which is leveraged by all client libraries is helpful in getting started, regardless of which Azure service you want to use.

```

vhvb1989 commented:
Since this is ReadMe only, no need to wait for CI

RickWinter commented on README.md:
```diff
@@ -66,42 +66,122 @@ In order to use the SDK installed via vcpkg with CMake, you can use the toolchai
+- Abstractions for Azure SDK credentials (`TokenCredential`).
+- Handling streaming data and input/output (I/O) via `BodyStream` along with its derived types.
+- Polling long-running operations (LROs), via `Operation<T>`.
+- Operations that return collections do so via a `PagedResponse<T>`.
```
```suggestion
- Collections are returned via `PagedResponse<T>`.

```

ahsonkhan commented:
Alright, since we are in the nit territory at this point, I am doing one last round of addressing of feedback, and I think I can merge it after. It's docs, we can always wordsmith and improve them post-GA (and I am sure we will) :)

antkmsft commented on README.md:
```diff
@@ -66,42 +66,122 @@ In order to use the SDK installed via vcpkg with CMake, you can use the toolchai
-std::string GetConnectionString();
+// Since they can contain secrets, get the required account name and key string from an environment
+// variable or Azure Key Vault. Leaving this up to the user to implement.
+std::string GetEndPointUrl();
```
```suggestion
std::string GetEndpointUrl();

```

antkmsft commented on README.md:
```diff
@@ -66,42 +66,122 @@ In order to use the SDK installed via vcpkg with CMake, you can use the toolchai
 {
-  std::string containerName = "testcontainer";
-  std::string blobName = "sample-blob";
+  std::string endpointUrl = GetEndPointUrl();
```
```suggestion
  std::string endpointUrl = GetEndPointUrl();
```suggestion
  std::string endpointUrl = GetEndpointUrl();

```
```

ahsonkhan commented:
/check-enforcer override

ahsonkhan commented:
/check-enforcer override
